extensions/jms/src/main/java/org/mule/extensions/jms/api/operation/MessageBuilder.java

```diff
@@ -0,0 +1,253 @@
+
+  /**
+   * the custom user properties that should be set to this {@link Message}
+   */
+  @Parameter
+  @Optional
+  @NullSafe
+  @Content
+  // TODO MULE-10901: Nullsafe bug
+  private Map<String, Object> properties = new HashMap<>();
+
+  /**
+   * Creates a {@link Message} based on the provided configurations
+   * @param jmsSupport the {@link JmsSupport} used to create the JMSReplyTo {@link Destination}
+   * @param session the current {@link Session}
+   * @param config the current {@link JmsProducerProperties}
+   * @return the {@link Message} created by the user
+   * @throws JMSException if an error occurs
+   */
+  Message build(JmsSupport jmsSupport, Session session, JmsConfig config)
+      throws JMSException {
+
+    config.getEncoding();
+
+    Message message = toMessage(content, session);
+
+    setJmsCorrelationIdHeader(message);
+    setJmsTypeHeader(config.getProducerConfig(), message);
+    setJmsReplyToHeader(jmsSupport, session, message, replyTo);
+
+    setJmsxProperties(message);
+    setUserProperties(message);
+
+    if (sendContentType) {
+      setContentTypeProperty(config, message);
+    }
+
+    return message;
+  }
+
+  private void setJmsReplyToHeader(JmsSupport jmsSupport, Session session, Message message, JmsDestination replyDestination) {
+    try {
+
+      // TODO NullSafe should not propagate default objects unless inner NullSafe is declared
+
+      if (replyDestination != null &&
+          !isBlank(replyDestination.getDestination())) {
+        Destination destination = jmsSupport.createDestination(session, replyDestination.getDestination(),
+                                                               replyDestination.getDestinationType().isTopic());
+        message.setJMSReplyTo(destination);
+      }
+    } catch (DestinationNotFoundException | JMSException e) {
+      LOGGER.error("Unable to set JMSReplyTo header: ", e);
+    }
+  }
+
+  private void setContentTypeProperty(JmsConfig config, Message message) {
+    try {
+      message.setStringProperty(CONTENT_TYPE_JMS_PROPERTY, resolveOverride(config.getContentType(), contentType));
+    } catch (JMSException e) {
+      LOGGER.error(format("Unable to set property [%s] of type String: ", CONTENT_TYPE_JMS_PROPERTY), e);
+    }
+  }
+
+  private void setJmsxProperties(final Message message) {
+    jmsxProperties.asMap().entrySet().stream()
+        .filter(e -> e.getValue() != null)
+        .forEach(e -> setJmsPropertySanitizeKeyIfNecessary(message, e.getKey(), e.getValue()));
+  }
+
+  private void setUserProperties(final Message message) {
+    properties.keySet().stream()
+        .filter(key -> !isBlank(key) && !JMSX_NAMES.contains(key))
+        .forEach(key -> setJmsPropertySanitizeKeyIfNecessary(message, key, properties.get(key)));
+  }
+
+
+  private void setJmsPropertySanitizeKeyIfNecessary(Message msg, String key, Object value) {
+    try {
+      // sanitize key as JMS Property Name
+      key = encodeHeader(key);
```

Comment:
you mean failing if the user-provided property name is not compliant with what JMS allows as key?